src/lib.ts

```diff
@@ -0,0 +1,30 @@
+import { setFailed } from "@actions/core";
+import { init, deploy, generateBranch } from "./git";
+import { action, actionInterface } from "./constants";
```

Comment:
When you import the module here it has a side effect where it initialises `action`.  Which still requires `FOLDER` so I get errors if I don't set the env variable. The initialisation code should be encapsulated in a function that takes in the initial object so that for example if I set just `{ folder: 'build', branch: 'gh-pages' }` the rest of the defaults are set. Currently this is not the case.

Reply:
I think the most recent change should be what you're looking for, let me know.

Reply:
@JamesIves I just tested it and I'm getting the same error, and `folder` is still marked as required.

Reply:
My bad, forgot to check that in. Should be there now!

Reply:
That works now. The only problem is the `repositoryPath` and `tokenType` are undefined because they aren't initialised with the settings I pass in.

Reply:
Those should all be functions now 👍

Reply:
I don't think it worked no matter what I do I get the following error.
```
::error::You must provide the action with either a Personal Access Token or the GitHub Token secret in order to deploy. If you wish to use an ssh deploy token then you must set SSH to true.
There was an error initializing the repository: Error: No deployment token/method was provided. You must provide the action with either a Personal Access Token or the GitHub Token secret in order to deploy. If you wish to use an ssh deploy token then you must set SSH to true.
Initialization step complete...
[command]/usr/local/bin/git ls-remote --heads undefined gh-pages | wc -l
```

### Configuration
```js
  const deployOptions: githubPagesDeploy.actionInterface = {
    gitHubRepository: core.getInput('REPOSITORY', {required: true}),
    branch: 'gh-pages',
    baseBranch: 'master',
    folder: 'book',
    root: '.',
    workspace: process.env['GITHUB_WORKSPACE']!
  }
```

Reply:
I think you need to prefix the environment variable with with `INPUT_`?  I'm guessing GitHub actions namespaces the variable when they are included as environment variable: https://github.com/actions/toolkit/blob/master/packages/core/src/core.ts#L67-L75. I'll give this a try myself today, it's been quite a hectic week so far so I haven't been able to devote a lot of time to ironing the bugs out on this just yet.

Reply:
Sorry, I have been passing `INPUT_GITHUB_TOKEN` and `INPUT_TOKEN` in the environment as well. It's fine if can't get to this immediately.

Reply:
Finally got around to testing this myself!

With the latest update, the following works for me;

```javascript
import run from 'github-pages-deploy-action';

const deployOptions = {
  gitHubRepository: 'JamesIves/github-pages-deploy-action',
  branch: 'gh-pages',
  folder: 'book',
  workspace: '/src',
}

run(deployOptions)
```

Reply:
Did you push those changes? I'm still getting this error, and with that configuration I get the following error.

```console
src/lib.ts:20:11 - error TS2741: Property 'root' is missing in type '{ gitHubRepository: string; branch: string; folder: string; workspace: string; }' but required in type 'actionInterface'.

20     const deployOptions: githubPagesDeploy.actionInterface = {
             ~~~~~~~~~~~~~

  ../github-pages-deploy-action/lib/constants.d.ts:32:5
    32     root: string;
           ~~~~
    'root' is declared here.

```

Reply:
That's weird - I guess not? They are 100% there now.

79ceb1e
3527f62

Reply:
Thanks! Also `lib.ts` doesn't re-export `run` is that intentional?

Reply:
Thanks! I also found a couple of bugs in the code. One being`git config` fails if you don't put the name in quotes, `run` doesn't stop when `git` produces an error, and `::error::(s || "").replace is not a function`.

```
[command]/usr/local/bin/git config user.name GitHub Pages Deploy Action
usage: git config [<options>]

Config file location
    --global              use global config file
    --system              use system config file
    --local               use repository config file
    -f, --file <file>     use given config file
    --blob <blob-id>      read config from given blob object

Action
    --get                 get value: name [value-regex]
    --get-all             get all values: key [value-regex]
    --get-regexp          get values for regexp: name-regex [value-regex]
    --get-urlmatch        get value specific for the URL: section[.var] URL
    --replace-all         replace all matching variables: name value [value_regex]
    --add                 add a new variable: name value
    --unset               remove a variable: name [value-regex]
    --unset-all           remove all matches: name [value-regex]
    --rename-section      rename section: old-name new-name
    --remove-section      remove a section: name
    -l, --list            list all
    -e, --edit            open an editor
    --get-color           find the color configured: slot [default]
    --get-colorbool       find the color setting: slot [stdout-is-tty]

Type
    -t, --type <>         value is given this type
    --bool                value is "true" or "false"
    --int                 value is decimal number
    --bool-or-int         value is --bool or --int
    --path                value is a path (file or directory name)
    --expiry-date         value is an expiry date

Other
    -z, --null            terminate values with NUL byte
    --name-only           show variable names only
    --includes            respect include directives on lookup
    --show-origin         show origin of config (file, standard input, blob, command line)
    --default <value>     with --get, use default value when missing entry

There was an error initializing the repository: Error: The process '/usr/local/bin/git' failed with exit code 129
Initialization step complete...
[command]/usr/local/bin/git ls-remote --heads https://c74315a9584bcb461b6a1c6b86829dd888e5d27c@github.com/XAMPPRocky/project-group-template.git gh-pages | wc -l
136f56c4b811f0906380fdeb35c722b5175238fc	refs/heads/gh-pages
[command]/usr/local/bin/git checkout --progress --force master
Already on 'master'
Your branch is up to date with 'origin/master'.
[command]/usr/local/bin/git fetch https://c74315a9584bcb461b6a1c6b86829dd888e5d27c@github.com/XAMPPRocky/project-group-template.git
From https://github.com/XAMPPRocky/project-group-template
 * branch            HEAD       -> FETCH_HEAD
[command]/usr/local/bin/git worktree add --checkout gh-action-temp-deployment-folder origin/gh-pages
Preparing worktree (detached HEAD 136f56c)
HEAD is now at 136f56c Deploy e232516ec16b01d1b80b94a4dfb92b61076a7319 to gh-pages
[command]/usr/bin/rsync -q -av --progress book/. gh-action-temp-deployment-folder --exclude .ssh --exclude .git --exclude .github
[command]/usr/local/bin/git status --porcelain
The deployment encountered an error. ❌
Completed Deployment ✅
::error::(s || "").replace is not a function
```

Reply:
Thanks! I also found a couple of bugs in the code. One being`git config` fails if you don't put the name in quotes, `run` doesn't stop when `git` produces an error, and `::error::(s || "").replace is not a function`.

```
[command]/usr/local/bin/git config user.name GitHub Pages Deploy Action
usage: git config [<options>]

Config file location
    --global              use global config file
    --system              use system config file
    --local               use repository config file
    -f, --file <file>     use given config file
    --blob <blob-id>      read config from given blob object

Action
    --get                 get value: name [value-regex]
    --get-all             get all values: key [value-regex]
    --get-regexp          get values for regexp: name-regex [value-regex]
    --get-urlmatch        get value specific for the URL: section[.var] URL
    --replace-all         replace all matching variables: name value [value_regex]
    --add                 add a new variable: name value
    --unset               remove a variable: name [value-regex]
    --unset-all           remove all matches: name [value-regex]
    --rename-section      rename section: old-name new-name
    --remove-section      remove a section: name
    -l, --list            list all
    -e, --edit            open an editor
    --get-color           find the color configured: slot [default]
    --get-colorbool       find the color setting: slot [stdout-is-tty]

Type
    -t, --type <>         value is given this type
    --bool                value is "true" or "false"
    --int                 value is decimal number
    --bool-or-int         value is --bool or --int
    --path                value is a path (file or directory name)
    --expiry-date         value is an expiry date

Other
    -z, --null            terminate values with NUL byte
    --name-only           show variable names only
    --includes            respect include directives on lookup
    --show-origin         show origin of config (file, standard input, blob, command line)
    --default <value>     with --get, use default value when missing entry

There was an error initializing the repository: Error: The process '/usr/local/bin/git' failed with exit code 129
Initialization step complete...
[command]/usr/local/bin/git ls-remote --heads https://<TOKEN>@github.com/XAMPPRocky/project-group-template.git gh-pages | wc -l
136f56c4b811f0906380fdeb35c722b5175238fc	refs/heads/gh-pages
[command]/usr/local/bin/git checkout --progress --force master
Already on 'master'
Your branch is up to date with 'origin/master'.
[command]/usr/local/bin/git fetch https://<TOKEN>@github.com/XAMPPRocky/project-group-template.git
From https://github.com/XAMPPRocky/project-group-template
 * branch            HEAD       -> FETCH_HEAD
[command]/usr/local/bin/git worktree add --checkout gh-action-temp-deployment-folder origin/gh-pages
Preparing worktree (detached HEAD 136f56c)
HEAD is now at 136f56c Deploy e232516ec16b01d1b80b94a4dfb92b61076a7319 to gh-pages
[command]/usr/bin/rsync -q -av --progress book/. gh-action-temp-deployment-folder --exclude .ssh --exclude .git --exclude .github
[command]/usr/local/bin/git status --porcelain
The deployment encountered an error. ❌
Completed Deployment ✅
::error::(s || "").replace is not a function
```

Reply:
Couple of changes.

* Fixed the issue with quotation marks being non-present throwing an error.
* Fixed the error handling, the action will now throw and setFailed from `run` instead of how it was done before (which admittedly was weird and not correct). This will cause the action to exit immediately after it encounters an error.
* Moved the credential and folder checking etc into a new function which gets called on every invocation. This is a more friendly approach for those using it as a package.
* I set `silence: true` on the `exec` command unless `INPUT_DEBUG` is set to `true`. This will prevent lines with the full git repository url from being exposed in the terminal. The console logs the action provides should be sufficient.
* Overall improved the tests and cases where errors are thrown. 
* Removed the need to set `accessToken` as an environment variable. It's still recommended but you can now simply hand in `accessToken: process.env['ACCESS_TOKEN']`, this combined with the hidden logs should prevent tokens from leaking. 

The following configuration is what I tested these changes with:

```javascript
import run from 'github-pages-deploy-action';

const deployOptions = {
  gitHubRepository: 'JamesIves/github-pages-deploy-action',
  branch: 'gh-pages',
  folder: 'book',
  workspace: 'src',
  accessToken: process.env['ACCESS_TOKEN']
}

run(deployOptions)
```

Reply:
Hey, this looks great, however I'm still getting the following error when running `run`.

```console
::error::(s || "").replace is not a function
```

Reply:
Do you have a log you can provide? This error is seemingly pretty generic.

Reply:
Aha, I've found the error. It's related to `setFailed` when it catches in the `run` function. I've pushed up a fix for this. You're probably getting another error though as this runs in the encompassing catch function.

Reply:
You're right, the error I get now is the following. The `git status —porcelain` command always seems to fail.

```
[command]/usr/local/bin/git status --porcelain
[command]/bin/rm -rf gh-action-temp-deployment-folder
::error::The deploy step encountered an error: There was an error when attempting to execute the process '/usr/local/bin/git'. This may indicate the process failed to start. Error: spawn /usr/local/bin/git ENOENT ❌
Deployment Failed ❌
```

Reply:
Are you using sub modules, and can you run this command outside of the context of the action? I can’t seem to replicate this issue.

Reply:
Oh sorry, I should have said. I'm running this locally on my Macbook. This is the script I use to test it and the configuration I run.

### test.sh
```bash
#!/usr/bin/env bash

git clone git@github.com:XAMPPRocky/project-group-template.git ./forge

GITHUB_WORKSPACE=./forge \
INPUT_TOKEN=$GITHUB_TOKEN \
INPUT_GITHUB_TOKEN=$GITHUB_TOKEN \
INPUT_ACCESS_TOKEN=$GITHUB_TOKEN \
DEBUG_DEPLOY_ACTION=1 \
INPUT_REPOSITORY=XAMPPRocky/project-group-template \
node lib/main.js

rm -rf ./forge
```

### actionInterface configuration
```javascript
{
      gitHubRepository: core.getInput('REPOSITORY', {required: true}),
      branch: 'gh-pages',
      folder: 'book',
      workspace: process.env['GITHUB_WORKSPACE'],
}
```

Reply:
Are you able to provide an example of this not working, minus the repository tokens? I've tried a numbers of ways to replicate this but it seems to be working for me 😞

Reply:
@XAMPPRocky - Found the issue and pushed up a fix for it. I used your repository as a bse, here was my config;

**test.sh**

```
#!/usr/bin/env bash
git clone git@github.com:JamesIves/test-project.git forge

GITHUB_WORKSPACE=./forge \
INPUT_ACCESS_TOKEN=$ACCESS_TOKEN \
DEBUG_DEPLOY_ACTION=1 \
INPUT_REPOSITORY=JamesIves/test-project \

node lib/main.js
```

**lib/main.js**

```javascript
const run = require('github-pages-deploy-action').default;

run({
  gitHubRepository: process.env['INPUT_REPOSITORY'],
  debug: true,
  branch: 'gh-pages',
  folder: 'draft-rfcs',
})
```

Reply:
Thank you it now works! Though I have a slightly different issue that might just be me misunderstanding how to use this library. That project is a [`mdbook`](https://github.com/rust-lang/mdBook) that lets you create static websites from markdown files. So I want to run `mdbook build` which builds all the files I want to deploy in the `book/` folder. Currently after building it says there's nothing to commit and stops early.

Reply:
That sounds correct, it will present this error if git cannot detect a diff between the files that get built into the folder and the files on the branch. Have you tried making a change so the compiled version has a change detected by git?

Reply:
Okay, I've checked and everything seems to work now! 🎉

Reply:
Woo! I'll get this merged and promoted.

Reply:
This is now released on npm! Just a heads up, I changed `gitHubRepository:` to `repositoryName:`. It still expects the same string, it just made more sense.

https://github.com/JamesIves/github-pages-deploy-action/releases/tag/3.3.1